### PR TITLE
Add D2Lang Unicode::strlen

### PR DESCRIPTION
**Pull request by IAmTrial**
These changes add the `Unicode::strlen` function. The function behaves similar to what is expected for a `strlen` function, with one notable exception. If the supplied string pointer is `NULL`, then the function returns 0.

When the function is compiled with Visual C++ 6.0, the binary is the same as 1.10f.

### CODE REVIEW

**Comment by Necrolis** [comment]
Is there any reason why this doesn't match the def of the demangled symbol?
`?strlen@Unicode@@SIHPBU1@@Z` demangles to `public: static int__fastcallUnicode::strlen(struct Unicode const *)`

that is it takes in a constant pointer to non-constant data, your def is taking in a non-constant pointer to constant data

**Comment by IAmTrial** [comment]
> Is there any reason why this doesn't match the def of the demangled symbol?
> `?strlen@Unicode@@SIHPBU1@@Z` demangles to `public: static int__fastcallUnicode::strlen(struct Unicode const *)`
> 
> that is it takes in a constant pointer to non-constant data, your def is taking in a non-constant pointer to constant data

[The type `const Unicode*` is the same as `Unicode const*`. The first declaration (the one I use) is "const West", while the second one is "East const" declaration style.](https://hackingcpp.com/cpp/design/east_vs_west_const.html) A const pointer to a `Unicode` would be `Unicode* const`, which has to be declared with "East const" style. If I change the type to `Unicode* const`, then the program doesn't link due to a missing definition.

**Comment by Necrolis** [comment]
Seems I misread things, so ignore my comment :)

**Review by Lectem — CHANGES_REQUESTED** [review]

File: source/D2Lang/src/D2Unicode/D2UnicodeStr.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include <D2Unicode.h>
+
+int __fastcall Unicode::strlen(const Unicode* str) {
+  int i = 0;
+
+  if (str == NULL) {
+    return i;
+  }
+
+  while (str[i].ch != L'\0') {
```
[the remark this answers is not in the document]
Comment (IAmTrial) [reply source/D2Lang/src/D2Unicode/D2UnicodeStr.cpp:34]: I agree that the results of the benchmark show a noticeable difference in performance when compiled for x86_64 targets, but disagree that the results would have the same impact in x86 (32-bit) targets. Using the exact same compiler and settings, with the exception of adding the `-m32` flag to enable targeting 32-bit x86, I found that the targets produced identical binaries. Results can be found here: https://godbolt.org/z/d5Ta83dn9

Despite this, the suggested change produces an overall positive with no significant drawbacks, so I'll agree to make the change.

**Pull request merged**